src/common/entity/entity_filter.ts

```diff
@@ -1,78 +1,114 @@
+import type { HassEntity } from "home-assistant-js-websocket";
+import type { HomeAssistant } from "../../types";
+import { ensureArray } from "../array/ensure-array";
 import { computeDomain } from "./compute_domain";
+import { getEntityContext } from "./context/get_entity_context";
 
-export type FilterFunc = (entityId: string) => boolean;
+type EntityCategory = "none" | "config" | "diagnostic";
 
 export interface EntityFilter {
-  include_domains: string[];
-  include_entities: string[];
-  exclude_domains: string[];
-  exclude_entities: string[];
+  domain?: string | string[];
+  device_class?: string | string[];
+  device?: string | string[];
+  area?: string | string[];
+  floor?: string | string[];
+  label?: string | string[];
+  entity_category?: EntityCategory | EntityCategory[];
+  hidden_platform?: string | string[];
 }
 
-export const isEmptyFilter = (filter: EntityFilter) =>
-  filter.include_domains.length +
-    filter.include_entities.length +
-    filter.exclude_domains.length +
-    filter.exclude_entities.length ===
-  0;
+type EntityFilterFunc = (entityId: string) => boolean;
 
-export const generateFilter = (
-  includeDomains?: string[],
-  includeEntities?: string[],
-  excludeDomains?: string[],
-  excludeEntities?: string[]
-): FilterFunc => {
-  const includeDomainsSet = new Set(includeDomains);
-  const includeEntitiesSet = new Set(includeEntities);
-  const excludeDomainsSet = new Set(excludeDomains);
-  const excludeEntitiesSet = new Set(excludeEntities);
+export const generateEntityFilter = (
+  hass: HomeAssistant,
+  filter: EntityFilter
+): EntityFilterFunc => {
+  const domains = new Set(ensureArray(filter.domain) ?? []);
+  const deviceClasses = new Set(ensureArray(filter.device_class) ?? []);
+  const floors = new Set(ensureArray(filter.floor) ?? []);
+  const areas = new Set(ensureArray(filter.area) ?? []);
+  const devices = new Set(ensureArray(filter.device) ?? []);
+  const entityCategories = new Set(ensureArray(filter.entity_category) ?? []);
+  const labels = new Set(ensureArray(filter.label) ?? []);
+  const hiddenPlatforms = new Set(ensureArray(filter.hidden_platform) ?? []);
 
-  const haveInclude = includeDomainsSet.size > 0 || includeEntitiesSet.size > 0;
-  const haveExclude = excludeDomainsSet.size > 0 || excludeEntitiesSet.size > 0;
+  return (entityId: string) => {
+    const stateObj = hass.states[entityId] as HassEntity | undefined;
+    if (!stateObj) {
+      return false;
+    }
+    if (domains.size > 0) {
```

Comment:
Wouldn't it make more sense to have domains set to `undefined` if it's to be tested?

Reply:
(same for all others)